55_156_159_Interpolator/corrected_interpolator.py

```diff
@@ -0,0 +1,162 @@
+                                         given_time2, value2):
+  #First determine the slope of the path between the two times. 
+  # This is the slope of the line known as a linear interpolant.
+  # The slope of a line where the horizontal axis is time,
+  # and the vertical axis is the value to interpolate, is the 
+  # rate of change in the value per unit time.
+  rate_of_change = (value2-value1) / (given_time2-given_time1)
+  # Difference between the time of interest and the fistTime
+  time_difference = time_of_interest-given_time1
+  # The interpolated data value is the early value plus the time difference 
+  # multiplied by the slope (change rate) 
+  interpolated_value = value1 + rate_of_change * time_difference;
+  # Print interpolated point to screen
+  print(f"[{time_of_interest}s {interpolated_value:.2f}K]. "
+        f"Interpolated from: [{given_time1:4.2f}s {value1:.2f}K] and "
+        f"[{given_time2:4.2f}s {value2:.2f}K]")
+  # The interpolated value is returned out of this function.
+  return interpolated_value;
+  
+# This function generates a fake time step for a fake sensor.
+def generate_fake_sensor_time(last_time): 
+  #Suppose the acquisition rate is about once every 0.5 seconds
+  acquisition_rate = 500 
+  #Suppose the time has a random 40 ms variation
+  acquisition_variation = random.randint(0, 40) 
+  # Here is computed the time step
+  acquisition_step_time = acquisition_rate + acquisition_variation 
+  # Convert to seconds from ms
+  acquisition_step_time = acquisition_step_time / 1000
+  time.sleep(acquisition_step_time) # Create wait time realistically 
+  #Returns the new time step which has random variation
+  return last_time + acquisition_step_time 
+ 
+# Generates a fake sensor reading. 
+# Maybe we can imagine this as the PPM of some substance being exhaled. 
+def generate_fake_sensor_reading(last_value):
+  generation_rate = 13
+  generation_variation = random.randint(0, 20)
+  generation_step = generation_rate + generation_variation
+  generation_step = generation_step / 100
+  return last_value + generation_step
+  
+# Generates a fake time+value pair from previous data point
+def generate_fake_pair(last_time, last_value):
+    fake_pair = []
+    fake_pair.append(generate_fake_sensor_time(last_time))
+    fake_pair.append(generate_fake_sensor_reading(last_value))
+    return fake_pair
+
+# Fake sensor is meant to be run in a thread to generate fake sensor data  
+def fake_sensor(name, start_time, start_value, time_limit, sensor_data) :
+  current_time = start_time
+  current_value = start_value
+  fakeSensorArguments = {"last_time":current_time, "last_value":current_value}
+  fake_pair = []
+  while current_time < time_limit :
+    fake_pair = fake_pair[:]
+    fake_pair = generate_fake_pair(**fakeSensorArguments)
+    current_time, current_value = fake_pair
+    fakeSensorArguments = {"last_time":current_time, "last_value":current_value}
+    sensor_data.append(fake_pair)
```

Comment:
The problem here with changing the passing in of the list to returning it, is that the sensor_data is passed in by reference and is being observed in real time by the other thread, so if it waited until the end to return the sensor data then no data would be gathered until the sensor had completely finished. Instead of being passed in by reference, it could have been a global variable.